aten/src/ATen/core/type.cpp

```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  types = this->types().vec();
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{types.begin(), types.end()};
+  std::copy_if(rhs_types.begin(), rhs_types.end(),
+            std::back_inserter(types),
+            [&](const TypePtr type) {
+              bool res = !dict.count(type);
+              dict.insert(type);
+              return res;
+            });
+  return types;
+}
+
+std::vector<TypePtr> UnionType::union_of(const UnionTypePtr rhs) const {
+  return union_of(rhs->types());
+}
+
+std::vector<TypePtr> UnionType::intersection_of(c10::ArrayRef<TypePtr> rhs_types) const {
+  std::vector<TypePtr> types;
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{this->types().begin(), this->types().end()};
+  for (auto type : rhs_types) {
+    if (dict.count(type)) {
+      types.emplace_back(type);
+    }
+  }
+  return types;
+}
+
+std::vector<TypePtr> UnionType::intersection_of(const UnionTypePtr rhs) const {
+  return intersection_of(rhs->types());
+}
+
+std::vector<TypePtr> UnionType::set_difference_of(c10::ArrayRef<TypePtr> rhs_types) const {
+  std::vector<TypePtr> types;
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{rhs_types.begin(), rhs_types.end()};
+  for (auto type : this->types()) {
+    if (!dict.count(type)) {
+      types.emplace_back(type);
+    }
+  }
+  return types;
+}
+
+std::vector<TypePtr> UnionType::set_difference_of(const UnionTypePtr rhs) const {
+  return intersection_of(rhs->types());
+}
+
+bool UnionType::is_optional_of_type(const TypePtr type) const {
+  TORCH_INTERNAL_ASSERT(type->kind() == OptionalType::Kind,
+                        "Expected Optional type, but found ", type->str());
+  if (!is_optional()) {
+    return false;
+  }
+  const auto inner = type->expect<OptionalType>()->getElementType();
+  // Same logic as `isSubtypeOf`, but we can't call that function with
+  // `this` as an argument
+  return std::any_of(this->types().begin(),
+                     this->types().end(),
+                     [&, this](TypePtr const union_contained_type){
+                        return inner->isSubtypeOf(union_contained_type);
+                      });
+}
+
+c10::optional<TypePtr> UnionType::to_optional() const {
+  if (!is_optional()) {
+      return c10::nullopt;
+  }
+  TypePtr contained = types_[0]->kind()!= NoneType::Kind ?
+                      types_[0] : types_[1];
+  return OptionalType::create(std::move(contained));
+}
+
+OptionalType::OptionalType(TypePtr contained)
+                           : UnionType({contained}, TypeKind::OptionalType) {
+  TORCH_INTERNAL_ASSERT(contained, "OptionalType requires a valid TypePtr");
+  std::vector<TypePtr> types;
+  flatten_dynamic_type(contained, types);
+  sort_and_filter(types);
+  if (types.size() == 1) {
+    contained_ = types[0];
+  }
+  else if (types.size() == 2) {
+    contained_ = types[0]->kind()!= NoneType::Kind ? types[0] : types[1];
+  }
+  else {
+    contained_ = UnionType::create(types);
```

Comment:
having a Optional[Union] seems wrong...

Reply:
This goes back to the problem we had with the `Union` constructor and factory function earlier...if I'm stuck returning `OptionalType`, what's the best thing to do when I actually want to return `Union[T1, ..., TN, None]`? Throw an error?